src/pudl_archiver/utils.py

```diff
@@ -0,0 +1,51 @@
+"""Utility functions used in multiple places within PUDL archiver."""
+import asyncio
+import logging
+from collections.abc import Awaitable, Callable
+from typing import Any
+
+import aiohttp
+
+logger = logging.getLogger(f"catalystcoop.{__name__}")
+
+
+async def retry_async(
+    async_func: Callable[..., Awaitable[Any]],
+    args: list | None = None,
+    kwargs: dict | None = None,
+    retry_count: int = 7,
+    retry_base_s: int = 2,
+    retry_on: tuple[type[Exception], ...] = (aiohttp.ClientError, asyncio.TimeoutError),
+):
+    """Retry a function that returns a coroutine, with exponential backoff.
+
+    Args:
+        async_func: the function to retry.
+        args: a list of args to pass in to the retried function.
+        kwargs: a dictionary of kwargs to pass into the retried function.
+        retry_count: how many total tries to make.
+        retry_base_s: how many seconds to wait the first time we retry.
+        retry_on: the Exception subclasses to retry on. Defaults to ClientError
+            and TimeoutError because that's what aiohttp.ClientSession.request
+            and the various .read()/.text() methods on aiohttp.ClientResponse
+            can throw.
+    """
+    if args is None:
+        args = []
+    if kwargs is None:
+        kwargs = {}
+    for try_count in range(1, retry_count + 1):
+        # try count is 1 indexed for logging clarity
+        coro = async_func(*args, **kwargs)
+        try:
+            return await coro
+        # aiohttp client can either throw ClientError or TimeoutError
```

Comment:
This comment is maybe unnecessary now that the exceptions to catch are configurable and this is explained in the docstring